### PR TITLE
Kill Soul class, new Cortex class

### PR DESCRIPTION
**Pull request by kafischer**
CortexDispatcher: Dispatcher registration and execution of MentalProcesses. The manager is essentially a converter between event driven and functional paradigms

A MentalProcess is:
- a container in which CortexSteps are taken. One version of CortexStep is returns and passed on the next one
- queued when received
- a queuing strategy for processes can be specified
- processes have an abort controller, which can be accessed in the queuing strategy

Goal is simplest possible api to convert from event driven arch to something that works with CortexStep

Trivial example:
```
const SamanthaReplies = async (
  signal: AbortSignal,
  newMemory: ChatMessage,
  lastStep: CortexStep
) => {
  let step = lastStep;
  step = step.withMemory([newMemory]);
  const shouts = await step.next(Action.EXTERNAL_DIALOG, {
    action: "shouts in all caps",
  });
  if (signal.aborted) {
    return step;
  } else {
    console.log("Samantha says: ", shouts.value);
    return shouts;
  }
};
const simpleQueuingStrategy: QueuingStrategy = (currentJob, queue, newJob) => {
  currentJob?.abortController?.abort();
  return [newJob];
};
const samanthaRepliesConfig: ProcessConfig = {
  name: "SamanthaReplies",
  process: SamanthaReplies,
};

const initialMemories = [
  {
    role: "system",
    content: "<Context>You are modeling the mind of Samantha</Context>",
  },
] as ChatMessage[];
let firstStep = new CortexStep("Samantha");
firstStep = firstStep.withMemory(initialMemories);
const cortex = new CortexDispatcher(firstStep, {
  queuingStrategy: simpleQueuingStrategy,
});
cortex.register(samanthaRepliesConfig);

// run
cortex.dispatch("SamanthaReplies", {
  role: "user",
  content: "Hello, Samantha!",
} as ChatMessage);

cortex.dispatch("SamanthaReplies", {
  role: "user",
  content: "F U!",
} as ChatMessage);
```

TODO
- [ ] delete the examples/integrations... sigh
- [ ] useSoul... ?
- [ ] READMEs
- [ ] move runnable examples to tests

TODO
- [ ] Perhaps Soul does live on in the C.AI style of creation, but just isn't really very configurable
- [ ] Add ids to memories to make them easy to change/update (imagine like a Turning machine almost operating on the memories)
- [ ] Something about updating a memory section or single memory based on the current understanding... like `withMemoryEffect`

### CODE REVIEW

**Comment by kafischer** [comment]
Hmm perhaps it should pass in an empty `CortexStep`?

Then would need the name in the `CortexManager` but could be interesting

**Comment by kafischer** [comment]
To demonstrate the desired pattern here

**Comment by kafischer** [comment]
Another possibility I kind of like would also be to return a CortexStep as input to the next event

**Comment by kafischer** [comment]
Hmm perhaps that is v0 CortexManager - totally functional and just in the space of CortexStep

**Comment by kafischer** [comment]
Makes an ridiculously simple api concept... just need to add naming and state mutations...

**Comment by kafischer** [comment]
Yea this is it! A turing complete framework for language model programming

**Comment by kafischer** [comment]
Can extend as our understanding becomes more advanced

**Comment by kafischer** [comment]
done

File: core/src/cortexManager.ts [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+  }
+
+  queueProcess(name: string, newMemory: ChatMessage) {
+    const process = this.processes.get(name);
+    if (!process) throw new Error(`Process ${name} does not exist`);
+
+    const job: Job = {
+      process,
+      newMemory,
+      abortController: new AbortController(),
```
Comment (tobowers) [inline core/src/cortexManager.ts:73]: how is this ever accessed?

File: core/src/integrations/nextjs.ts [hunk 2]
```diff
@@ -189,96 +181,96 @@ export type UIThought = {
...
-
-      if (soulStartsConversation) {
-        setTimeout(() => conversation.current.tell(""), 200);
-      }
-    }
-  }, []);
-
-  return { tellSoul, soulReads, messages, soulThoughts };
-}
+// export type UseSoulConfig = {
```
Comment (tobowers) [inline core/src/integrations/nextjs.ts:184]: just delete all this?

File: core/src/cortexManager.ts [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+    };
+
+    this.processQueue = this.queuingStrategy(
+      this.currentJob,
+      this.processQueue,
+      job
+    );
+    if (!this.isDispatching) {
```
Comment (tobowers) [inline core/src/cortexManager.ts:81]: because dispatch() is async I *think* there's a race condition here.

File: core/src/cortexManager.ts [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+  private isDispatching = false;
+
+  constructor(firstStep: CortexStep, options?: ManagerOptions) {
+    if (options?.queuingStrategy) {
+      this.queuingStrategy = options.queuingStrategy;
+    }
+    this.lastStep = firstStep;
+  }
+
+  registerProcess({ name, process }: ProcessConfig) {
```
Comment (tobowers) [inline core/src/cortexManager.ts:62]: "process" is a bit confusing, but after reading the interfaces, now I get it.

File: core/src/blueprint.ts [hunk 5]
```diff
@@ -1,16 +1,10 @@
...
-  Introspective,
-  ReflectiveLP,
-}
-
 export interface Blueprint {
   name: string;
   essence: string;
   personality?: string;
   initialPlan?: string;
-  thoughtFramework?: ThoughtFramework;
```
Comment (tobowers) [inline core/src/blueprint.ts:13]: 👍

**Review by tobowers — COMMENTED** [review]
If it's helpful, here's a class I use in all my projects: https://gist.github.com/tobowers/1dce59028ab39a8d2ca02b1c45e4c9e3

Looks good! I just had a couple of questions.

File: core/src/cortexManager.ts (hunk 3 above)
Comment (kafischer) [reply core/src/cortexManager.ts:81]: in fact there are many... oops

File: core/src/cortexManager.ts (hunk 1 above)
Comment (kafischer) [reply core/src/cortexManager.ts]: Only ever in the Queuing strategy method

e.g.
```
const abortQueuingStrategy: QueuingStrategy = (currentJob, queue, newJob) => {
  currentJob?.abortController?.abort();
  return [newJob];
};
```

File: core/src/cortexManager.ts (hunk 4 above)
Comment (kafischer) [reply core/src/cortexManager.ts]: is there a better name?

File: core/src/integrations/nextjs.ts (hunk 2 above)
Comment (kafischer) [reply core/src/integrations/nextjs.ts]: actually i wonder if this and Soul lives on, but as a class that's just not very configurable

**Comment by kafischer** [comment]
Thinking more on this - I don't think I'm going to kill the Soul class/code, but maybe will remove some of the configurability and keep it as an intro to the project

**Comment by kafischer** [comment]
Merged in the CortexScheduler in #69

**Pull request closed without merging**